Aikido.Zen.Core/Helpers/RateLimitingHelper.cs

```diff
@@ -27,36 +32,115 @@ public static bool IsAllowed(string key, int windowSizeInMS, int maxRequests)
             if (windowSizeInMS <= 0) return true;
 
             var currentTime = GetCurrentTimestamp();
-            RequestInfo requestInfo;
 
-            // since http requests are handled in parallel, we need to lock the cache to prevent race conditions
             lock (_lock)
             {
-                if (!RateLimitedItems.TryGetValue(key, out requestInfo))
+                // Get or create the list of timestamps for this key
+                if (!RateLimitedItems.TryGetValue(key, out var timestamps))
                 {
-                    RateLimitedItems.Set(key, new RequestInfo(1, currentTime));
+                    timestamps = new List<long> { currentTime };
+                    RateLimitedItems.Set(key, timestamps);
                     return true;
                 }
 
-                var elapsedTime = currentTime - requestInfo.StartTime;
+                // Remove timestamps that are outside the window
+                timestamps.RemoveAll(timestamp => currentTime - timestamp > windowSizeInMS);
 
-                if (elapsedTime >= windowSizeInMS)
+                // Add current timestamp
+                timestamps.Add(currentTime);
+
+                // Update the cache with filtered timestamps
+                RateLimitedItems.Set(key, timestamps);
+
+                // Check if the number of requests is within limits
+                return timestamps.Count <= maxRequests;
+            }
+        }
+
+        /// <summary>
+        /// Checks if a route matches a wildcard pattern
+        /// </summary>
+        /// <param name="routePattern">The route pattern that may contain wildcards (*)</param>
+        /// <param name="actualRoute">The actual route to check against the pattern</param>
+        /// <returns>True if the route matches the pattern, false otherwise</returns>
+        public static bool IsWildcardMatch(string routePattern, string actualRoute)
```

Comment:
This logic should not be in Ratelimiting classes, this is also used to match for stuff like allowed ips, forced-off protection, ..., would create a MatchingEndpoints helper or something akin to that